wdl/gwas/regenie_step1.wdl

```diff
@@ -12,13 +12,74 @@ task step1 {
     String options
 
     String docker
-
+    Int covariate_inclusion_threshold
     command <<<
 
         set -euxo pipefail
 
-        n_cpu=`grep -c ^processor /proc/cpuinfo`
 
+        #filter degenerate covariates, e.g. sex imputed from sex-specific phenotypes
+        COVARFILE=${cov_pheno}
+        PHENO="${phenolist sep=','}"
+        COVARS="${covariates}"
+        THRESHOLD=${covariate_inclusion_threshold}
+        # Filter binary covariates that don't have enough covariate values in them
+        # Inputs: covariate file, comma-separated phenolist, comma-separated covariate list, threshold for excluding a covariate
+        # If a covariate is quantitative (=having values different from 0,1,NA), it is masked and will be passed through.
+        # If a binary covariate has value NA, it will not be counted towards 0 or 1 for that covariate.
+        # If a covariate does not seem to exist (e.g. PC{1:10}), it will be passed through.
+        # If any of the phenotypes is not NA on that row, that row will be counted. This is in line with the step1 mean-imputation for multiple phenotypes.
+        zcat $COVARFILE |awk -v covariates=$COVARS  -v phenos=$PHENO -v th=$THRESHOLD > new_covars  '
+        BEGIN{FS="\t"}
+        NR == 1 {
+            covlen = split(covariates,covars,",")
+            phlen = split(phenos,phenoarr,",")
+            for (i=1; i<=NF; i++){
+                h[$i] = i
+                mask[$i] = 0
+            }
+        }
+        NR > 1 {
+            #if any of the phenotypes is not NA, then take the row into account
+            process_row=0
+            for (ph in phenoarr){
+                if ($h[phenoarr[ph]] != "NA"){
+                    process_row = 1
+                }
+            }
+            if (process_row == 1){
+                for (co in covars){
+                    if($h[covars[co]] == 0) {
+                        zerovals[covars[co]] +=1
+                    }
+                    else if($h[covars[co]] == 1) {
+                        onevals[covars[co]] +=1
+                    }
+                    else if($h[covars[co]] == "NA"){
+                        #no-op
+                        na=0;
+                    }
+                    else {
+                        #mask this covariate to be included, no matter the counts
+                        #includes both covariate not found in header and quantitative covars
+                        mask[covars[co]] =1
+                    }
+                }
+            }
+
+        }
+        END{
+            SEP=""
+            for (co in covars){
+                if( ( zerovals[covars[co]] > th && onevals[covars[co]] > th ) || mask[covars[co]] == 1 ){
+                    printf("%s%s" ,SEP,covars[co])
+                    SEP=","
+                }
```

Comment:
could print to stderr the covar name and counts of 1/0 indeed here but great otherwise!!!